xdem/coreg/affine.py

```diff
@@ -740,282 +782,1289 @@ def vertical_shift(
     return vshift, subsample_final
 
 
-##################################
-# Affine coregistration subclasses
-##################################
+############################
+# 4/ Iterative closest point
+############################
 
-AffineCoregType = TypeVar("AffineCoregType", bound="AffineCoreg")
 
+def _icp_fit_func(
+    inputs: tuple[NDArrayf, NDArrayf, NDArrayf | None],
+    t1: float,
+    t2: float,
+    t3: float,
+    alpha1: float,
+    alpha2: float,
+    alpha3: float,
+    method: Literal["point-to-point", "point-to-plane"],
+) -> NDArrayf:
+    """
+    Fit function of ICP, a rigid transformation with 6 parameters (3 translations and 3 rotations) between closest
+    points (that are fixed for the optimization, and update at each iterative step).
+
+    :param inputs: Constant input for the fit function: three arrays of size 3xN, for the reference point cloud, the
+        to-be-aligned point cloud and the plane normals.
+    :param t1: Translation in X.
+    :param t2: Translation in Y.
+    :param t3: Translation in Z.
+    :param alpha1: Rotation around X.
+    :param alpha2: Rotation around Y.
+    :param alpha3: Rotation around Z.
+    :param method: Method of iterative closest point registration, either "point-to-point" of Besl and McKay (1992)
+        that minimizes 3D distances, or "point-to-plane" of Chen and Medioni (1992) that minimizes 3D distances
+        projected on normals.
+
+    :return Array of distances between closest points.
+    """
+
+    # Get inputs
+    ref, tba, norm = inputs
 
-class AffineCoreg(Coreg):
+    # Build an affine matrix for 3D translations and rotations
+    matrix = matrix_from_translations_rotations(t1, t2, t3, alpha1, alpha2, alpha3, use_degrees=False)
+
+    # Apply affine transformation
+    trans_tba = _apply_matrix_pts_mat(mat=tba, matrix=matrix)
+
+    # Define residuals depending on type of ICP method
+    # Point-to-point is simply the difference, from Besl and McKay (1992), https://doi.org/10.1117/12.57955
+    if method == "point-to-point":
+        diffs = (trans_tba - ref) ** 2
+    # Point-to-plane used the normals, from Chen and Medioni (1992), https://doi.org/10.1016/0262-8856(92)90066-C
+    # A priori, this method is faster based on Rusinkiewicz and Levoy (2001), https://doi.org/10.1109/IM.2001.924423
+    elif method == "point-to-plane":
+        assert norm is not None
+        diffs = (trans_tba - ref) * norm
+    else:
+        raise ValueError("ICP method must be 'point-to-point' or 'point-to-plane'.")
```

Comment:
Ideally this should be caught at the beginning of the function

Reply:
Done, this was a duplicate check.